akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -137,6 +144,129 @@ object GraphStageLogic {
   private object DoNothing extends (() ⇒ Unit) {
     def apply(): Unit = ()
   }
+
+  trait MinimalDeathWatch {
+    this: MinimalActorRef ⇒
+    import StageActorRef._
+    private[this] val _watchedBy = new AtomicReference[Set[ActorRef]](ActorCell.emptyActorRefSet)
+
+    def log: LoggingAdapter
+
+    override def isTerminated = _watchedBy.get() == StageTerminatedTombstone
+
+    //noinspection EmptyCheck
+    protected def sendTerminated(): Unit = {
+      val watchedBy = _watchedBy.getAndSet(StageTerminatedTombstone)
+      val wasAlreadyTerminated = watchedBy == StageTerminatedTombstone
```

Comment:
I'm not sure I see your point, it's a normal nullcheck, because the Tombstone could have been set already. Or we extend the thinking about "it is only ever called from stop(), from beforePostStop() which is safe", then sure - no need to nullcheck. Shall I remove it?